Move to adaptor backend

nathanjmcdougall opened the pull request:
To set us up for #263 and #254.

At @machow's suggestion here: https://github.com/rstudio/pins-python/pull/263#issuecomment-2259174212
I definitely agree this is definitely a nicer way of doing things.

nathanjmcdougall commented on pins/boards.py:
```diff
@@ -1147,46 +1154,17 @@ def prepare_pin_version(self, pin_dir_path, x, name: "str | None", *args, **kwar
 
-        # data preview ----
-
-        # TODO: move out data_preview logic? Can we draw some limits here?
```
Addressing this TODO; this logic has been moved to `_DFAdaptor` and `_Adaptor` respectively.

nathanjmcdougall commented on pins/tests/_databackend/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Technically complying with MIT requires distributing the license

nathanjmcdougall replied:
Actually this probably need to be moved out of `tests` because I'm not sure if that gets included in the wheel build (I think it doesn't).

isabelizimm replied:
The tests directory does get included in the wheel build! I'm not entirely sure where this file should be, but it might be easier to not vendor in this package, if possible 😄

nathanjmcdougall commented on pins/_adaptors.py:
```diff
@@ -0,0 +1,172 @@
+from abc import abstractmethod
+from typing import TYPE_CHECKING, Any, ClassVar, TypeAlias, overload
+
+from typing_extensions import Self
```
I'll need to avoid using `Self`. I forget that `typing_extensions` isn't a built-in...

isabelizimm commented:
I would prefer to not vendor in all of `databackend`; @machow, wdyt of putting the latest GH release on PyPI? It looks like there is just v0.0.1, some sort of [auth-y error for the CI releasing 0.0.2](https://github.com/machow/databackend/actions/runs/5212817528)?

isabelizimm commented:
Okay, we should be able to use [databackend==0.0.3](https://pypi.org/project/databackend/#history) now, rather than vendoring it all in 🎉

nathanjmcdougall commented:
Cool, I've rebased away the commit with the vendored code and added the new dependency.

isabelizimm commented:
paired with @machow, he'll be taking over this PR for review/next steps!

nathanjmcdougall commented:
@machow hey, it would be good if you could review this PR sometime. No great rush but I'd really love to get #249.

machow commented on pins/_adaptors.py:
```diff
@@ -0,0 +1,172 @@
+    def write_json(self, file: str) -> None: ...
+    @overload
+    def write_json(self, file: None) -> str: ...
+    def write_json(self, file: str | None = None) -> str | None:
```
Okay I think I understand -- this looks like an override of the original `.write_json()` method, but its job seems different (it's used for the data preview). Its job looks more like the added `.shape` or `.columns` properties.

Can we do this?:

* rename `.write_json()` here to something else (maybe `.to_json()`)
* change the parent annotation for `write_json()` to always return None (not sometimes a str)

nathanjmcdougall replied:
See what I've done in this commit:

https://github.com/rstudio/pins-python/pull/298/commits/daa4239ed4f7b32e3fc93caf8817f2f3f65ab221

machow commented on pins/drivers.py:
```diff
@@ -135,6 +129,8 @@ def save_data(
     #       as argument to board, and then type dispatchers for explicit cases
     #       of saving / loading objects different ways.
 
+    adaptor = _create_adaptor(obj)
```
Could we allow `obj` to be either `_Adaptor | Any` (the typing is redundant, but maybe a helpful signal)? Then, if obj is not an adaptor, this line could create one.

I'm guessing keeping the original `save_data` behavior is useful for testing, but it'd be nice not to have to roundtrip by calling it on `save_data(_Adaptor._d, ...)` in this board method:

https://github.com/rstudio/pins-python/pull/298/files#diff-36792b1eedbe5453d2c6b58286ab65eed0c9286e94e2695a736edff37164e885R757

nathanjmcdougall replied:
I see you've done that here:

https://github.com/rstudio/pins-python/pull/298/commits/d0fa9c96d9ad762ba4fd6bc28d68ee75802922f6

I will:
- Write a test case for this
- Refactor to take advantage of this

machow commented on pins/drivers.py:
```diff
@@ -149,39 +145,22 @@ def save_data(
-        _assert_is_pandas_df(obj, file_type=type)
-
-        raise NotImplementedError(
+        msg = (
```
~Can we put messages back into their exceptions, rather than defining an intermediate variable? (I could be missing context where it was decided this should happen, so I could be totally off base here).~

It looks like defining a `msg = ` variable is common in libraries like polars, so punting this comment 😅. If you know of a place where this approach is mentioned, might come from, I'd be curious to here -- since it'll be handy to point to!

machow replied:
Oh, interesting -- I hadn't seen the error message lint rules before. I'm onboard with whatever seems most useful to folks. I wonder if these checks are disabled by default, and the select rule in pyproject.toml might have enabled them?

Opened issue #323 for tracking!

machow reviewed (changes requested):
I'm so sorry for the long wait 😓 -- this is looking really good! I think the 3 small things I thought could benefit from a little tweaking are..

* It seems like _Adaptor._d should not be typed as a class variable
* `_Adaptor.write_json()` is overriden in `_DfAdaptor` in a surprising way (it returns a string), in order to work for data previews. I wonder if the `_DFAdaptor` version could be renamed to `.to_json()` to reflect its doing a different job.
* `save_data()` is fed an object pulled from `_Adaptor._d` but then re-creates the adaptor. Could we allow `save_data()` to take an adaptor, to prevent this round tripping?

Thanks again for all the work you put into this! If it's useful, I'm happy to pick this up and finish -- especially since so much time has passed, and I've picked up the context again / probably owe it to folks 😭

nathanjmcdougall commented:
I'll take a look at this later today - but I'm happy for you to pick it up if you prefer.

nathanjmcdougall commented on pins/_adaptors.py:
```diff
@@ -0,0 +1,172 @@
+    _DataFrame: TypeAlias = _PandasDataFrame
+
+
+class _AbstractPandasFrame(AbstractBackend):
```
Yeah I agree with that.

nathanjmcdougall commented on pins/_adaptors.py:
```diff
@@ -0,0 +1,172 @@
+
+
+class _Adaptor:
+    _d: ClassVar[Any]
```
Yeah also agreed. I can't recall why I did that, it might have been some misguided thoughts about pyright behaviour.

nathanjmcdougall commented on pins/_adaptors.py:
```diff
@@ -0,0 +1,172 @@
+        self._d = data
+
+    @overload
+    def write_json(self, file: str) -> None: ...
```
Totally, good idea.

machow commented:
Ah, we're both pushing to this PR -- I'll leave it to you, thanks for working on this! I noticed that a doctest test fails now, and seems to be related to a more recent release of fsspec.

I'm guessing the ffspec github backend doesn't like the two `//` in this path anymore: `pins/tests/pins-compat/df_csv/20220214T163720Z-9bfad//df_csv.csv`

If it's useful, we can always ignore that for now, and handle as a separate from this PR

nathanjmcdougall commented:
Sorry yeah unfortunately I think we both started working on it at the same time.

Sounds good - I'm in favour of fixing it in another PR first so after I've addressed the issues here I'll work on that.

machow reviewed (approved):
I only took a quick glance over the most recent changes -- but this LGTM (isabel feel free to look closer!)

isabelizimm reviewed (approved):
This looks good to me! I added in `typing_extensions` as a dependency, which can be dropped after 3.9 isn't supported.
